File: backend/rust-clean-coke-architecture-patterns/templates/usecase.rs
// Template: replace CreateExampleEntity*, ExampleEntity*, and ExampleRepository
// with project-specific names. Usecases do not import Axum, Diesel, schema, or row types.

use std::sync::Arc;

use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::domain::entities::ExampleEntity;
use crate::domain::repositories::ExampleRepository;
use crate::domain::value_objects::ExampleEntityName;
use crate::usecases::UsecaseError;

pub struct CreateExampleEntityInput {
    pub owner_id: Uuid,
    pub column_text: String,
    pub column_url: String,
}

pub struct CreateExampleEntityOutput {
    pub id: Uuid,
    pub owner_id: Uuid,
    pub column_text: String,
    pub column_url: String,
    pub status: String,
    pub created_at: DateTime<Utc>,
}

pub struct CreateExampleEntityUseCase {
    example_repo: Arc<dyn ExampleRepository>,
}

impl CreateExampleEntityUseCase {
    pub fn new(example_repo: Arc<dyn ExampleRepository>) -> Self {
        Self { example_repo }
    }

    pub async fn execute(
        &self,
        input: CreateExampleEntityInput,
    ) -> Result<CreateExampleEntityOutput, UsecaseError> {
        let column_text = ExampleEntityName::new(input.column_text)?;

        let entity = ExampleEntity::new(input.owner_id, column_text, input.column_url);

        self.example_repo.create(&entity).await?;

        Ok(CreateExampleEntityOutput {
            id: *entity.id().as_uuid(),
            owner_id: *entity.owner_id(),
            column_text: entity.column_text().as_str().to_string(),
            column_url: entity.column_url().to_string(),
            status: entity.status().as_str().to_string(),
            created_at: entity.created_at(),
        })
    }
}

File: plugins/eng/skills/go-clean-architecture/templates/usecase.go
// Template: replace CreateExampleEntity*, ExampleEntity*, and examplefeature
// with project-specific names. Target location:
// internal/usecase/{feature}/{action}.go — package named after the feature.
// Replace "app" in imports with the project's module path.
// Usecases do not import Fiber, pgx, sqlc-generated packages, or handler DTOs.
//
// This usecase has no special opinion about any repository failure, so every
// error goes through usecase.ConvertError. When a usecase does have an opinion
// — an ErrNotFound that means "then create it", a unique violation on an
// idempotency key that means "already processed" — match it with errors.Is
// first and keep ConvertError as the fallback for the rest. Nothing raw leaves
// Execute, and a usecase logs only errors it handles itself.
package examplefeature

import (
	"context"
	"time"

	"github.com/google/uuid"

	"app/internal/domain/entity"
	"app/internal/domain/repository"
	"app/internal/domain/valueobject"
	"app/internal/usecase"
)

type CreateExampleEntityInput struct {
	OwnerID uuid.UUID
	Name    string
	URL     string
}

type CreateExampleEntityOutput struct {
	ID        uuid.UUID
	OwnerID   uuid.UUID
	Name      string
	URL       string
	Status    string
	CreatedAt time.Time
}

type CreateExampleEntityUsecase struct {
	exampleRepo repository.ExampleRepository
}

func NewCreateExampleEntityUsecase(exampleRepo repository.ExampleRepository) *CreateExampleEntityUsecase {
	return &CreateExampleEntityUsecase{exampleRepo: exampleRepo}
}

func (u *CreateExampleEntityUsecase) Execute(
	ctx context.Context,
	input CreateExampleEntityInput,
) (*CreateExampleEntityOutput, error) {
	name, err := valueobject.NewExampleEntityName(input.Name)
	if err != nil {
		return nil, usecase.ConvertError(err)
	}

	e := entity.NewExampleEntity(input.OwnerID, name, input.URL)

	if err := u.exampleRepo.Create(ctx, e); err != nil {
		return nil, usecase.ConvertError(err)
	}

	return &CreateExampleEntityOutput{
		ID:        e.ID().UUID(),
		OwnerID:   e.OwnerID(),
		Name:      e.Name().String(),
		URL:       e.URL(),
		Status:    e.Status().String(),
		CreatedAt: e.CreatedAt(),
	}, nil
}
